File: src/semp_workflow/modules/client_username.py
"""Client Username module - client_username.add, client_username.delete."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient
from ..semp.helpers import check_name_length, clean_payload, coerce_bool, enc
from .base import BaseModule

logger = logging.getLogger(__name__)


def _build_username_payload(args: dict) -> dict:
    payload = clean_payload(args)
    if "enabled" in payload:
        payload["enabled"] = coerce_bool(payload["enabled"])
    return payload


class ClientUsernameAdd(BaseModule):
    description = "Create a client username on the message VPN. Skipped if the username already exists."
    params = {
        "clientUsername":    {"type": "string",  "required": True,  "description": "The client username to create"},
        "clientProfileName": {"type": "string",  "required": False, "description": "Client profile to assign to this username", "default": "default"},
        "aclProfileName":    {"type": "string",  "required": False, "description": "ACL profile to assign to this username", "default": "default"},
        "password":          {"type": "string",  "required": False, "description": "Password for the client username"},
        "enabled":           {"type": "boolean", "required": False, "description": "Enable the client username after creation", "default": "true"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        username = args.get("clientUsername", "")
        if not username:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: clientUsername")

        if err := check_name_length("clientUsername", username):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"clientUsernames/{enc(username)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking client username: {e}")

        if exists:
            return ActionResult(
                ResultStatus.SKIPPED, f"Client username '{username}' already exists"
            )

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would create client username '{username}'")

        try:
            payload = _build_username_payload(args)
            client.create("clientUsernames", payload)
            return ActionResult(
                ResultStatus.OK,
                f"Client username '{username}' created",            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED,
                f"Failed to create client username '{username}': {e}",
            )


class ClientUsernameDelete(BaseModule):
    description = "Delete a client username from the message VPN. Skipped if the username does not exist."
    params = {
        "clientUsername": {"type": "string", "required": True, "description": "The client username to delete"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        username = args.get("clientUsername", "")
        if not username:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: clientUsername")

        if err := check_name_length("clientUsername", username):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"clientUsernames/{enc(username)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking client username: {e}")

        if not exists:
            return ActionResult(
                ResultStatus.SKIPPED, f"Client username '{username}' does not exist"
            )

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would delete client username '{username}'")

        try:
            client.delete(path)
            return ActionResult(
                ResultStatus.OK,
                f"Client username '{username}' deleted",            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED,
                f"Failed to delete client username '{username}': {e}",
            )


MODULES = {
    "client_username.add": ClientUsernameAdd,
    "client_username.delete": ClientUsernameDelete,
}

File: internal/modules/client_username.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("client_username.add", clientUsernameAdd{})
	register("client_username.delete", clientUsernameDelete{})
	register("client_username.update", clientUsernameUpdate{})
}

// buildUsernamePayload cleans args and coerces the enabled flag.
func buildUsernamePayload(args map[string]any) map[string]any {
	payload := semp.CleanPayload(args)
	coerceBoolFields(payload, "enabled")
	return payload
}

type clientUsernameAdd struct{}

func (clientUsernameAdd) Description() string {
	return "Create a client username on the message VPN. Skipped if the username already exists."
}

func (clientUsernameAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "clientUsername", Type: "string", Required: true, Description: "The client username to create"},
		{Name: "clientProfileName", Type: "string", Description: "Client profile to assign to this username", Default: "default"},
		{Name: "aclProfileName", Type: "string", Description: "ACL profile to assign to this username", Default: "default"},
		{Name: "password", Type: "string", Description: "Password for the client username"},
		{Name: "enabled", Type: "boolean", Description: "Enable the client username after creation", Default: "true"},
	}
}

func (clientUsernameAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	username := argStr(args, "clientUsername", "")
	if username == "" {
		return failed("Missing required arg: clientUsername")
	}
	if err := semp.CheckNameLength("clientUsername", username); err != "" {
		return failed(err)
	}

	path := "clientUsernames/" + semp.Enc(username)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking client username: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("Client username '%s' already exists", username))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would create client username '%s'", username))
	}

	if _, err := c.Create("clientUsernames", buildUsernamePayload(args)); err != nil {
		return failed(fmt.Sprintf("Failed to create client username '%s': %s", username, err))
	}
	return ok(fmt.Sprintf("Client username '%s' created", username))
}

type clientUsernameDelete struct{}

func (clientUsernameDelete) Description() string {
	return "Delete a client username from the message VPN. Skipped if the username does not exist."
}

func (clientUsernameDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "clientUsername", Type: "string", Required: true, Description: "The client username to delete"},
	}
}

func (clientUsernameDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	username := argStr(args, "clientUsername", "")
	if username == "" {
		return failed("Missing required arg: clientUsername")
	}
	if err := semp.CheckNameLength("clientUsername", username); err != "" {
		return failed(err)
	}

	path := "clientUsernames/" + semp.Enc(username)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking client username: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("Client username '%s' does not exist", username))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would delete client username '%s'", username))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to delete client username '%s': %s", username, err))
	}
	return ok(fmt.Sprintf("Client username '%s' deleted", username))
}

type clientUsernameUpdate struct{}

func (clientUsernameUpdate) Description() string {
	return "Update attributes of an existing client username. Fails if the username does not exist."
}

func (clientUsernameUpdate) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "clientUsername", Type: "string", Required: true, Description: "The client username to update"},
		{Name: "clientProfileName", Type: "string", Description: "Client profile to assign"},
		{Name: "aclProfileName", Type: "string", Description: "ACL profile to assign"},
		{Name: "password", Type: "string", Description: "Password for the client username"},
		{Name: "enabled", Type: "boolean", Description: "Enable or disable the client username"},
	}
}

func (clientUsernameUpdate) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	username := argStr(args, "clientUsername", "")
	if username == "" {
		return failed("Missing required arg: clientUsername")
	}
	if err := semp.CheckNameLength("clientUsername", username); err != "" {
		return failed(err)
	}

	path := "clientUsernames/" + semp.Enc(username)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking client username: %s", err))
	}
	if !exists {
		return failed(fmt.Sprintf("Client username '%s' does not exist", username))
	}

	payload := buildUsernamePayload(args)
	delete(payload, "clientUsername")

	if len(payload) == 0 {
		return skipped("No fields to update")
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would update client username '%s'", username))
	}

	if _, err := c.Update(path, payload); err != nil {
		return failed(fmt.Sprintf("Failed to update client username '%s': %s", username, err))
	}
	return ok(fmt.Sprintf("Client username '%s' updated", username))
}
